test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -0,0 +1,364 @@
+  return std::move(v);
+}
+
+/*
+ * chiSquareTest uses assert_matches_quantiles to check random numbers generated
+ * by a given callable generate_samples against the reference quantiles computed
+ * with the given callable generate_quantiles.
+ *
+ * The generate_samples callable must have the signature:
+ *   T_out generate_samples(T_param1 p1, T_param2 p2, T_rng)
+ * where T_param1 and T_param2 can be any combination of:
+ *   double, std::vector<double>, VectorXd, or RowVectorXd
+ * and T_rng is a boost random number generator
+ *
+ * The output of generate_samples must be of
+ *   stan::length(out) == stan::max_size(p1, p2)
+ *
+ * Each element of the output, out[i], should correspond to a random variate
+ * generated according to parameters p1[i] and p2[i]
+ *
+ * The generate_quantiles callable must have the signature:
+ *   std::vector<double> generate_samples(int N, double p1, double p2)
+ *
+ * N is the number of samples that the quantiles will be compared
+ * against and p1 and p2 the values of the parameters that these quantiles
+ * correspond to.
+ *
+ * The output of generate_quantiles should be the quantiles used for comparing
+ * empirical distributions used by assert_matches_quantiles
+ *
+ * Parameters for each of these callables are generated according to constraints
+ * c1 and c2.
+ *
+ * @tparam T_param1 Type of first parameter
+ * @tparam T_param2 Type of second parameter
+ * @tparam T_generate_samples Type of callable for generating samples
+ * @tparam T_generate_quantiles Type of callable for generating quantiles
+ * @param sampler sampler functor to test
+ * @param c1 Constraint of first parameter
+ * @param c2 Constraint of second parameter
+ */
+template<typename T_param1, typename T_param2,
+         typename T_generate_samples, typename T_generate_quantiles>
+void chiSquareTest(int N, int M_vec,
+                   T_generate_samples generate_samples,
+                   T_generate_quantiles generate_quantiles,
+                   Constraint c1,
+                   Constraint c2) {
+  boost::random::mt19937 rng;
+  T_param1 p1;
+  T_param2 p2;
+  int M = stan::max_size(M_vec, M_vec);
+  generateValues(p1, c1, M);
+  generateValues(p2, c2, M);
+  stan::scalar_seq_view<T_param1> p1_vec(p1);
+  stan::scalar_seq_view<T_param2> p2_vec(p2);
+
+  std::vector<std::vector<double> > samples_to_test_transpose;
+  for (int n = 0; n < N; ++n) {
+    // If p1 and p2 are scalars, the output is a scalar. Need to promote it
+    // to a std::vector
+    samples_to_test_transpose.push_back(
+      promote_to_vector(generate_samples(p1, p2, rng))
+    );
+  }
+  for (int m = 0; m < M; ++m) {
+    std::vector<double> samples_to_test;
+    for (int n = 0; n < N; ++n) {
+      samples_to_test.push_back(samples_to_test_transpose[n][m]);
+    }
+    std::vector<double> quantiles = generate_quantiles(N, p1_vec[m], p2_vec[m]);
+
+    assert_matches_quantiles(samples_to_test, quantiles, 1e-6);
```

Comment:
This is all about calling this function in the end. I think I should change the name of the test function to mirror this (the chiSquare name -- I guess I just copied and pasted it whatever haha)